### PR TITLE
Fix multiple issues with Object.defineProperty

### LINKED ISSUE #6358: Issue with "Object.defineProperty()": (How to remove getter or setter correctly)
MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty

#### What information was incorrect, unhelpful, or incomplete?
By default, get and set are `undefined`.
What value should I use if I want to keep current getter/setter? What value should I use if I want to delete anyone of them?

#### Specific section or headline?
Modifying a property

#### What did you expect to see?
What value should I use if I want to keep current getter/setter?
What value should I use if I want to delete anyone of them?

#### Did you test this? If so, how?
No, I didn't.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/object/defineproperty`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.html
* Last commit: https://github.com/mdn/content/commit/05b35ae2b888c9d1067affaa60af7010bc586a3d
* Document last modified: 2021-06-14T21:14:08.000Z

</details>

### LINKED ISSUE #8942: Issue with "Object.defineProperty()": More clarity needed for configuring non-configurable data property
MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty

#### What information was incorrect, unhelpful, or incomplete?

> It is not possible to change any attribute of a **non-configurable** accessor property. _For data properties which are **configurable**, it is possible to modify the value if the property is writable, and it is possible to change the `writable` attribute from `true` to `false`._

Obviously we can change its `value` and `writable` since it's _configurable_, after all. I think it was supposed to describe a situation where the data property is non-configurable. The sentences before and after it are both describing such cases -- about non-configurable properties.

#### Specific section or headline?

[Modifying a property](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty#modifying_a_property)

#### What did you expect to see?

> For data properties which are **non-configurable**, it is possible to modify the value if the property is writable, and it is possible to change the `writable` attribute from `true` to `false`.

#### Did you test this? If so, how?

Modified `value` and `writable` attribute of a non-configurable data property with `Object.defineProperty()`.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/object/defineproperty`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md
* Last commit: https://github.com/mdn/content/commit/cf607d68522cd35ee7670782d3ee3a361eaef2e4
* Document last modified: 2021-07-20T18:47:57.000Z

</details>

### LINKED ISSUE #15565: Object descriptor - configurable
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty

### What specific section or headline is this issue about?

Configurable attribute

### What information was incorrect, unhelpful, or incomplete?

As I see, you mean that **value** and **writable** flags could be changed even if **configurable** is set to false.
I tested it in chrome and **writable** can't be changed  - see screenshot.
image
May be there is some misunderstanding?

### What did you expect to see?

Falsy **configurable**  doesn't allow to change any flag

### Do you have any supporting links, references, or citations?

The configurable attribute controls at the same time whether the property can be deleted from the object and whether its attributes (other than value and writable) can be changed.

### Do you have anything more you want to share?

image

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/object/defineproperty`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md
* Last commit: https://github.com/mdn/content/commit/edaa4aff4cf7532efdcba6014b9316aad2a5b6f1
* Document last modified: 2022-03-27T17:30:40.000Z

</details>

### PR DESCRIPTION
**Pull request by Josh-Cena**
#### Motivation

Fix #15565
Fix #8942
Fix #6358

`Object.defineProperty` is so hard to get right >.<

#### Metadata

- [ ] Adds a new document
- [ ] Rewrites (or significantly expands) a document
- [x] Fixes a typo, bug, or other error

### CODE REVIEW

File: files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md [hunk 1]
```diff
@@ -57,8 +57,10 @@ Both data and accessor descriptors are objects. They share the following optiona
 properties using `Object.defineProperty()`):
 
 - `configurable`
-  - : `true` if the type of this property descriptor may be changed and if the
-    property may be deleted from the corresponding object.
+  - : when `false`:
+    - the type of this property cannot be changed between data property and accessor property;
+    - the property may not be deleted;
+    - other attributes of its descriptor cannot be changed (however, if it's a data descriptor with `writable: true`, the `value` can be changed, and `writable` can be changed to `false`);
```
Comment (teoli2003) [inline files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:63]: I find this a bit unclear. Are these conditions _or_ed or _and_ed ?

File: files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md [hunk 2]
```diff
@@ -144,6 +146,12 @@ Object.defineProperty(obj, 'key', withValue('static'));
 (Object.freeze || Object)(Object.prototype);
 ```
 
+When the property already exists, `Object.defineProperty()` attempts to modify the property according to the values in the descriptor and the property's current configuration.
+
+If the old descriptor had its `configurable` attribute set to `false`, the property is said to be "non-configurable". It is not possible to change any attribute of a non-configurable accessor property, and it is not possible to switch between data and accessor property types. For data properties with `writable: true`, it is possible to modify the value and change the `writable` attribute from `true` to `false`. A {{jsxref("TypeError")}} is thrown when attempts are made to change non-configurable property attributes (except `value` and `writable`, if permitted), except when defining a value same as the original value on a data property.
```
Comment (teoli2003) [inline files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:151]: Matter of personal taste:
```suggestion
If the old descriptor had its `configurable` attribute set to `false`, the property is said to be _non-configurable_. It is not possible to change an attribute of a non-configurable accessor property, and it is not possible to switch between data and accessor property types. For data properties with `writable: true`, it is possible to modify the value and change the `writable` attribute from `true` to `false`. A {{jsxref("TypeError")}} is thrown when attempts are made to change non-configurable property attributes (except `value` and `writable`, if permitted), except when defining a value same as the original value on a data property.
```

File: files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md [hunk 3]
```diff
@@ -195,19 +203,7 @@ Object.defineProperty(o, 'conflict', {
...
-change any attribute of a non-configurable accessor property. For data properties which are configurable, it is
-possible to modify the value if the property is writable, and it is possible to change the
-`writable` attribute from `true` to `false`. It is not
-possible to switch between data and accessor property types when the property is
-non-configurable.
-
-A {{jsxref("TypeError")}} is thrown when attempts are made to change non-configurable
-property attributes (except `value` and `writable`, if permitted)
-unless the current and new values are the same.
+When modifying an existing property, the current property configuration determines if the operator succeeds, does nothing, or throws a `TypeError`.
```
Comment (teoli2003) [inline files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:206]: New section, I would use a link here:
```suggestion
When modifying an existing property, the current property configuration determines if the operator succeeds, does nothing, or throws a {{jsxref("TypeError")}}.
```

**Review by teoli2003 — CHANGES_REQUESTED** [review]
Thanks a lot. One thing isn't clear to me and I think we should improve this part. + a few nits.

File: files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md (hunk 1 above)
Comment (Josh-Cena) [reply files/en-us/web/javascript/reference/global_objects/object/defineproperty/index.md:63]: They are the consequences of `configurable` being false, so "and" seems to be the only reasonable interpretation.

**Review by teoli2003 — APPROVED** [review]
Thank you!

**Pull request merged**